### PR TITLE
poll on FolderBranch in SubMan for empty TLF

### PR DESCRIPTION
**Pull request by songgao**
When user goes into an empty TLF in GUI and upload something, it doesn't show any changes because the subscription manager can't register for changes with an empty FolderBranch. This PR fixes that by polling on trying to get FolderBranch for the TLF at a 1s interval while subscription is active, and make the subscription actually happen and send a notification as soon as it gets a FolderBranch.

### CODE REVIEW

File: go/kbfs/libkbfs/subscription_manager.go [hunk 1]
```diff
@@ -315,18 +310,127 @@ func (sm *subscriptionManager) SubscribePath(ctx context.Context,
...
+	ps.subscriptionIDs[req.sid] = req.topic
+	ps.pathsToNotify[req.path] = struct{}{}
 
-	sm.pathSubscriptionIDToRef[sid] = ref
+	sm.pathSubscriptionIDToRef[req.sid] = ref
 	subscriptionIDSetter()
 	return nil
 }
 
+const folderBranchPollingInterval = time.Second
```
Comment (strib) [inline go/kbfs/libkbfs/subscription_manager.go:321]: Once this is merged with the other PR, it would be nice to have a `const` section at the top.

File: go/kbfs/libkbfs/subscription_manager.go [hunk 2]
```diff
@@ -315,18 +310,127 @@ func (sm *subscriptionManager) SubscribePath(ctx context.Context,
...
+				_ = sm.cancelAndDeleteFolderBranchPoller(req.sid)
+				return
+			}
+
+			if fb == (data.FolderBranch{}) {
+				continue
+			}
+
+			// We have a folderBranch now! Go ahead and complete the
+			// sbuscription, and send a notification too.
```
Comment (strib) [inline go/kbfs/libkbfs/subscription_manager.go:357]: "sbuscription" -> "subscription"

File: go/kbfs/libkbfs/subscription_manager.go [hunk 3]
```diff
@@ -315,18 +310,127 @@ func (sm *subscriptionManager) SubscribePath(ctx context.Context,
...
+	if err != nil {
+		return err
+	}
+	req := subscribePathRequest{
+		sid:                 sid,
+		path:                path,
+		topic:               topic,
+		deduplicateInterval: deduplicateInterval,
+	}
+	if fb != (data.FolderBranch{}) {
```
Comment (strib) [inline go/kbfs/libkbfs/subscription_manager.go:419]: Seems like we don't need to take the lock until right before this `if`, is that right?  No need to hold it when parsing or talking to FBO.

File: go/kbfs/libkbfs/subscription_manager.go [hunk 4]
```diff
@@ -315,18 +310,127 @@ func (sm *subscriptionManager) SubscribePath(ctx context.Context,
...
+	sm.subscribePathWithoutFolderBranchLocked(req, parsedPath)
+	return nil
+}
+
 // SubscribeNonPath implements the SubscriptionManager interface.
 func (sm *subscriptionManager) SubscribeNonPath(
 	ctx context.Context, sid SubscriptionID, topic keybase1.SubscriptionTopic,
 	deduplicateInterval *time.Duration) error {
+	// Lock at the beginnning to protect against racing with unsubscribe. We
+	// could still endup lingering subscription if unsubscribe happens to fast
```
Comment (strib) [inline go/kbfs/libkbfs/subscription_manager.go:431]: "endup lingering" -> "end up with a lingering"

"to fast" -> 'too fast"

File: go/kbfs/libkbfs/subscription_manager.go [hunk 5]
```diff
@@ -315,18 +310,127 @@ func (sm *subscriptionManager) SubscribePath(ctx context.Context,
...
+	}
+	sm.subscribePathWithoutFolderBranchLocked(req, parsedPath)
+	return nil
+}
+
 // SubscribeNonPath implements the SubscriptionManager interface.
 func (sm *subscriptionManager) SubscribeNonPath(
 	ctx context.Context, sid SubscriptionID, topic keybase1.SubscriptionTopic,
 	deduplicateInterval *time.Duration) error {
+	// Lock at the beginnning to protect against racing with unsubscribe. We
```
Comment (strib) [inline go/kbfs/libkbfs/subscription_manager.go:430]: This comment is unrelated to the rest of this PR right?  That's fine, just making sure I understand.

File: go/kbfs/libkbfs/subscription_manager.go [hunk 6]
```diff
@@ -361,6 +465,12 @@ func (sm *subscriptionManager) SubscribeNonPath(
 
 func (sm *subscriptionManager) unsubscribePathLocked(
 	ctx context.Context, subscriptionID SubscriptionID) {
+	// First check if this is a subscription we don't yet have a folderBranch
+	// for.
+	if sm.cancelAndDeleteFolderBranchPollerLocked(subscriptionID) {
```
Comment (strib) [inline go/kbfs/libkbfs/subscription_manager.go:470]: This path is taken during `Shutdown` right?  So we shouldn't have to worry about lingerings I think, just wanted to make sure.

File: go/kbfs/libkbfs/subscription_manager.go (hunk 5 above)
Comment (songgao) [reply go/kbfs/libkbfs/subscription_manager.go:430]: This is for this PR, but is supposed to be in SubscribePath instead of here, sorry! I'll explain in the other comment about locking.

File: go/kbfs/libkbfs/subscription_manager.go (hunk 3 above)
Comment (songgao) [reply go/kbfs/libkbfs/subscription_manager.go:419]: I intentionally put it at the beginning. (The other comment in SubscribeNonPath was for this). It's specifically to protect a race where unsubscribe happens immediately after subscribe, and we end up calling unsubscribePath before these happen, and that'd be a no-op. At the point we start the polling thing, unsubscribe has already returned and there won't be any more unsubscribe calls in the future for this subscriptionID, then we end up with a subscription that shouldn't be there.

File: go/kbfs/libkbfs/subscription_manager.go (hunk 6 above)
Comment (songgao) [reply go/kbfs/libkbfs/subscription_manager.go:470]: Yes this path is taken at shutdown. All subscriptions are gone at shutdown. The "lingering" thing was referring to the subscribe/unsubscribe race.

File: go/kbfs/libkbfs/subscription_manager.go (hunk 3 above)
Comment (strib) [reply go/kbfs/libkbfs/subscription_manager.go:432]: I guess the specific worry is that the `getFolderBranch()` call would take a non-zero amount of time in some cases?  Everything else should be pretty fast and the lock placement shouldn't matter at all, but I guess if the subscribe is actually creating the FBO (not sure if that can actually happen in production, since presumably the GUI would have done something to access the TLF before subscribing to it?).  I'm also slightly worried about a deadlock here if the FBO needs to call back into the subscriber somehow, maybe due to something we add in the future.  I guess I don't have a simple alternative suggestion though.

But maybe we could move the lock to just above the `getFolderBranch()` call in that case?  I'm just thinking that it will be clearer to have it protecting only the things that matter.  I guess I don't feel too strongly about it.

**Comment by songgao** [comment]
CI passed before rebase. Merging
